orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -885,23 +919,58 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
       .def(py::init([](const std::string& model_uri,
                        onnxruntime::training::api::CheckpointState* state,
                        std::optional<std::string> eval_model_uri,
-                       OrtDevice device) {
-        onnxruntime::SessionOptions session_option;
+                       OrtDevice device, PySessionOptions* session_options) {
         std::vector<std::shared_ptr<IExecutionProvider>> provider = GetExecutionProvidersForTrainingApis(device);
-
         auto env = GetTrainingEnv().GetORTEnv();
         return std::make_unique<onnxruntime::training::api::Module>(
-            model_uri, state, session_option, *env, provider, eval_model_uri);
+            model_uri, state, session_options->value, *env, provider, eval_model_uri,
+            session_options->custom_op_domains_);
       }))
       .def("train_step",
            [](onnxruntime::training::api::Module* model,
-              const std::vector<OrtValue>& user_inputs, std::vector<OrtValue>& user_outputs) -> void {
-             ORT_THROW_IF_ERROR(model->TrainStep(user_inputs, user_outputs));
+              const std::vector<py::object>& user_inputs, std::vector<OrtValue>& user_outputs) -> void {
+             std::vector<OrtValue> feeds;
+             const auto model_inputs_with_error = model->GetTrainingModelInputs();
+             ORT_THROW_IF_ERROR(model_inputs_with_error.first);
+             ORT_ENFORCE(model_inputs_with_error.second, "Training model graph inputs are not defined.");
+             for (size_t idx = 0; idx < user_inputs.size(); ++idx) {
+               auto& feed = user_inputs[idx];
+               const auto feed_name = model->GetTrainingModelInputName(idx);
+               // No need to process 'None's sent in by the user
+               // to feed Optional inputs in the graph.
+               // We just won't include anything in the feed and ORT
+               // will handle such implicit 'None's internally.
+               if (!feed.is(py::none())) {
+                 OrtValue ort_value;
+                 CreateGenericMLValue(model_inputs_with_error.second, GetAllocator(), feed_name, feed, &ort_value);
```

Comment:
hmmm, then it will be CPU only input.

Reply:
While the input is CPU only, the graph can be executed on cuda if the execution provider is cuda.

This is the primary difference between run with io binding and run in the onnxruntime inference realm.

But you're right. We should also allow users to execute the graph with inputs on gpu. Let me try to support both scenarios in this PR.